File: Languages/c/algorithms-dsa/leetcode/arrays-hashing/contains_duplicate.c
/**
 * LeetCode 217: Contains Duplicate
 * Given an integer array nums, return 1 if any value appears at least twice.
 *
 * Time: O(n^2) - simple approach without hash set
 * Space: O(1)
 */

#include <stdio.h>
#include <assert.h>

int contains_duplicate(int* nums, int nums_size) {
    for (int i = 0; i < nums_size; i++) {
        for (int j = i + 1; j < nums_size; j++) {
            if (nums[i] == nums[j]) {
                return 1;
            }
        }
    }
    return 0;
}

int main(void) {
    int nums1[] = {1, 2, 3, 1};
    assert(contains_duplicate(nums1, 4) == 1);

    int nums2[] = {1, 2, 3, 4};
    assert(contains_duplicate(nums2, 4) == 0);

    int nums3[] = {1, 1, 1, 3, 3, 4, 3, 2, 4, 2};
    assert(contains_duplicate(nums3, 10) == 1);

    // Empty array
    assert(contains_duplicate(NULL, 0) == 0);

    printf("All tests passed!\n");
    return 0;
}

File: Languages/go/algorithms-dsa/leetcode/arrays-hashing/contains_duplicate.go
/*
LeetCode #217: Contains Duplicate
Topic: Arrays & Hashing
Difficulty: Easy

Given an integer array nums, return true if any value appears at
least twice in the array, and return false if every element is distinct.

Approach: Use a hash set to track seen numbers. If we encounter a
number already in the set, return true.

Time: O(n), Space: O(n)
*/
package main

import "fmt"

func containsDuplicate(nums []int) bool {
	seen := make(map[int]bool)
	for _, num := range nums {
		if seen[num] {
			return true
		}
		seen[num] = true
	}
	return false
}

func main() {
	if containsDuplicate([]int{1, 2, 3, 1}) != true {
		fmt.Println("FAIL Test 1")
	} else {
		fmt.Println("PASS Test 1")
	}

	if containsDuplicate([]int{1, 2, 3, 4}) != false {
		fmt.Println("FAIL Test 2")
	} else {
		fmt.Println("PASS Test 2")
	}

	if containsDuplicate([]int{1, 1, 1, 3, 3, 4, 3, 2, 4, 2}) != true {
		fmt.Println("FAIL Test 3")
	} else {
		fmt.Println("PASS Test 3")
	}
}
